design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,288 @@
+
+All variable values need to be wrapped using single quotes. All variables are stored as string values in memory, and when
+replaced, the value without the quotes will be used. ksqlDB parsing will be in charge of determining the format at that
+time.
+
+The reason not to remove single-quotes from values is to allow spaces, and to leave the syntax open for using data types in the future if required.  
+Also, having decide the format at run-time (i.e. `define n = 3.34238`) is tricky as we could lose precision in case of decimals. Better stay with Strings for now.
+
+Define a new variable:
+```
+ksql> DEFINE replicas = '3';
+ksql> DEFINE partitions = '5';
+```
+
+Print a defined variable value:
+```
+ksql> DEFINE replicas;
+DEFINE replicas   = '3'
+```
+
+List all defined variables:
+```
+ksql> DEFINE;
+DEFINE replicas   = '3'
+DEFINE partitions = '5'
+```
+
+#### Deleting substitution variables
+
+The `UNDEFINE` command deletes the variable.
+
+Syntax:
+```
+UNDEFINE name;
+```
+
+### 2. Scope
+
+All variables will live during the session of a ksqlDB connection, then destroyed when the connection is closed. Only the user
+in that session can interact with the defined substitution variables. No other users can interact with other users' variables.
+
+In CLI, the session is opened when starting the CLI with the `ksql` command, and destroyed when typing `exit`.
+
+In headless, the session is opened when starting the ksqlDB server with a SQL script. The variables stay alive until the server is
+killed.
+
+Variables are not persisted anywhere. They are living in memory as long as the session is not terminated.
+
+#### RUN SCRIPT command
+
+If a `RUN SCRIPT` command is executed in a CLI session, then any variable defined there will be local to the execution context of the
+script. The variables will override any variable previously defined in the session. When the script is completed, all local variables
+defined will be destroyed. This ensures users do not override their session variables by accident when running `RUN SCRIPT` commands.
+
+#### The SOURCE command
+
+To allow users load variables and keep their scope in the session, then the new `SOURCE` command will be added. This will override
+any previously defined variables, and will be destroyed until the user leaves the session.
+
+Syntax:
+```
+SOURCE '<script_file>';
+```
+
+The `SOURCE` command will work like a `RUN SCRIPT`. If other SQL statements are in the file, then those statements will also be
+executed, thus to avoid limiting users from using DDL/DML commands while defining variables at the same time from one single script.
+
+`SOURCE` will be used only on the CLI and headless mode. REST API will not accept it because files may be out of the scope of the server filesystem.
+
+### 3. Referencing substitution variables
+
+Variables will be referenced by wrapping the variable between `{}` characters (i.e. `{replicas}`). These characters are also used by
```

Comment:
I might change the prefix/suffix for '${}' to avoid conflicting with JSON literal strings.

Reply:
+1 for `${}`.

Reply:
Done